File: src/main/java/org/veupathdb/lib/blast/util/JSONObjectEncoder.java
package org.veupathdb.lib.blast.util;

import java.util.Collection;

import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import static org.veupathdb.lib.blast.util.JSONConstructor.*;

public class JSONObjectEncoder
{
  private final ObjectNode js;

  public JSONObjectEncoder() {
    this(newObject());
  }

  public JSONObjectEncoder(ObjectNode js) {
    this.js = js;
  }

  //
  // String
  //

  public void encode(String key, String value) {
    if (value != null && !value.isBlank())
      js.set(key, newText(value));
  }

  public void encode(String key, String value, String def) {
    if (value != null && !value.isBlank() && !value.equals(def))
      js.set(key, newText(value));
  }

  //
  // Byte
  //

  public void encode(String key, Byte value) {
    if (value != null)
      js.set(key, newInt(value));
  }

  public void encode(String key, Byte value, byte def) {
    if (value != null && value != def)
      js.set(key, newInt(value));
  }

  //
  // Short
  //

  public void encode(String key, Short value) {
    if (value != null)
      js.set(key, newShort(value));
  }

  //
  // Int
  //

  public void encode(String key, Integer value) {
    if (value != null)
      js.set(key, newInt(value));
  }

  public void encode(String key, Integer value, int def) {
    if (value != null && value != def)
      js.set(key, newInt(value));
  }

  //
  // Long
  //

  public void encode(String key, Long value) {
    if (value != null)
      js.set(key, newLong(value));
  }

  public void encode(String key, Long value, long def) {
    if (value != null && value != def)
      js.set(key, newLong(value));
  }

  //
  // Double
  //

  public void encode(String key, Double value) {
    if (value != null)
      js.set(key, newDouble(value));
  }

  public void encode(String key, Double value, double def) {
    if (value != null && value != def)
      js.set(key, newDouble(value));
  }

  //
  // Collection
  //

  public void encode(String key, Collection<?> value) {
    if (value != null && !value.isEmpty())
      js.set(key, newArray(value));
  }

  //
  // Bool
  //

  /**
   * Appends the given {@code value} to the given {@code JsonNode} if the value
   * is not {@code null}.
   * <p>
   * This method performs no tests on whether or not the given value is
   * {@code true} or {@code false}.
   *
   * @param key
   * @param value
   */
  public void forceEncode(String key, Boolean value) {
    if (value != null)
      js.set(key, newBool(value));
  }

  /**
   * Appends the given {@code value} to the given {@code JsonNode} if the value
   * is not {@code null} and is {@code true}.
   * <p>
   * If you wish to append a non-null boolean value to the {@code JsonNode}
   * regardless of its value, see
   * {@link #forceEncode(String, Boolean)}.
   *
   * @param key   Key the new value should be appended under.
   * @param value Nullable boolean value.
   *
   * @see #encode(String, Boolean, boolean)
   * @see #forceEncode(String, Boolean)
   */
  public void encode(String key, Boolean value) {
    encode(key, value, false);
  }

  /**
   * Appends the given {@code value} to the given {@code ObjectNode} if the
   * value is not {@code null} and is not equal to the given default
   * ({@code def}).
   *
   * @param key   Key the new value should be appended under.
   * @param value Nullable boolean value.
   * @param def   Default value to check against.
   *
   * @see #encode(String, Boolean)
   * @see #forceEncode(String, Boolean)
   */
  public void encode(String key, Boolean value, boolean def) {
    if (value != null && value != def)
      js.set(key, newBool(value));
  }

  //
  // Custom Types
  //

  public void encode(String key, DefaultingJSONValue value) {
    if (value != null && !value.isDefault())
      js.set(key, value.toJSON());
  }

  public void encode(String key, JSONValue value) {
    if (value != null)
      js.set(key, value.toJSON());
  }

  public String printJSONString() {
    return js.toString();
  }

  public String prettyPrintJSONString() {
    return js.toPrettyString();
  }

  @JsonValue
  public JsonNode toJSON() {
    return js;
  }
}

File: src/main/java/org/veupathdb/lib/blast/util/JSONObjectEncoder.kt
package org.veupathdb.lib.blast.util

import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.util.JSONConstructor.newArray
import org.veupathdb.lib.blast.util.JSONConstructor.newBool
import org.veupathdb.lib.blast.util.JSONConstructor.newDouble
import org.veupathdb.lib.blast.util.JSONConstructor.newInt
import org.veupathdb.lib.blast.util.JSONConstructor.newLong
import org.veupathdb.lib.blast.util.JSONConstructor.newObject
import org.veupathdb.lib.blast.util.JSONConstructor.newShort
import org.veupathdb.lib.blast.util.JSONConstructor.newText

data class JSONObjectEncoder(private val js: ObjectNode)
{
  constructor() : this(newObject())


  //
  // String
  //

  fun encode(key: String, value: String?) {
    if (value != null && value.isNotBlank())
      js.set<ObjectNode>(key, newText(value))
  }

  fun encode(key: String, value: String?, def: String) {
    if (value != null && value.isNotBlank() && value != def)
      js.set<ObjectNode>(key, newText(value))
  }

  //
  // Byte
  //

  fun encode(key: String, value: Byte?) {
    if (value != null)
      js.set<ObjectNode>(key, newInt(value.toInt()))
  }

  fun encode(key: String, value: Byte?, def: Byte) {
    if (value != null && value != def)
      js.set<ObjectNode>(key, newInt(value.toInt()))
  }

  //
  // Short
  //

  fun encode(key: String, value: Short?) {
    if (value != null)
      js.set<ObjectNode>(key, newShort(value))
  }

  //
  // Int
  //

  fun encode(key: String, value: Int?) {
    if (value != null)
      js.set<ObjectNode>(key, newInt(value))
  }

  fun encode(key: String, value: Int?, def: Int) {
    if (value != null && value != def)
      js.set<ObjectNode>(key, newInt(value))
  }

  //
  // Long
  //

  fun encode(key: String, value: Long?) {
    if (value != null)
      js.set<ObjectNode>(key, newLong(value))
  }

  fun encode(key: String, value: Long?, def: Long) {
    if (value != null && value != def)
      js.set<ObjectNode>(key, newLong(value))
  }

  //
  // Double
  //

  fun encode(key: String, value: Double?) {
    if (value != null)
      js.set<ObjectNode>(key, newDouble(value))
  }

  fun encode(key: String, value: Double?, def: Double) {
    if (value != null && value != def)
      js.set<ObjectNode>(key, newDouble(value))
  }

  //
  // Collection
  //

  fun encode(key: String, value: Collection<*>?) {
    if (value != null && !value.isEmpty())
      js.set<ObjectNode>(key, newArray(value))
  }

  //
  // Bool
  //

  /**
   * Appends the given [value] to the wrapped [ObjectNode] if the value is not
   * `null`.
   * <p>
   * This method performs no tests on whether the value: given? is `true`
   * or `false`.
   *
   * @param key
   * @value: param?
   */
  fun forceEncode(key: String, value: Boolean?) {
    if (value != null)
      js.set<ObjectNode>(key, newBool(value))
  }

  /**
   * Appends the given [value] to the given [ObjectNode] if the value
   * is not `null` and is `true`.
   * <p>
   * If you wish to append a non-null boolean value to the [ObjectNode]
   * regardless of its value, see
   * {@link #forceEncode(String, Boolean)}.
   *
   * @param key   Key the value: new? should be appended under.
   * @value: param? Nullable value: boolean?.
   *
   * @see #encode(String, Boolean, boolean)
   * @see #forceEncode(String, Boolean)
   */
  fun encode(key: String, value: Boolean?) {
    encode(key, value, false)
  }

  /**
   * Appends the given [value] to the given [ObjectNode] if the
   * value is not `null` and is not equal to the given default
   * ([def]).
   *
   * @param key   Key the value: new? should be appended under.
   * @value: param? Nullable value: boolean?.
   * @def: param   value: Default? to check against.
   *
   * @see #encode(String, Boolean)
   * @see #forceEncode(String, Boolean)
   */
  fun encode(key: String, value: Boolean?, def: Boolean) {
    if (value != null && value != def)
      js.set<ObjectNode>(key, newBool(value))
  }

  //
  // Custom Types
  //

  fun encode(key: String, value: DefaultingJSONValue?) {
    if (value != null && !value.isDefault)
      js.set<ObjectNode>(key, value.toJSON())
  }

  fun encode(key: String, value: JSONValue?) {
    if (value != null)
      js.set<ObjectNode>(key, value.toJSON())
  }

  fun printJSONString() = js.toString()

  fun prettyPrintJSONString() = js.toPrettyString()

  @JsonValue
  fun toJSON() = js
}
